wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h

```diff
@@ -118,50 +130,56 @@ int esp_CryptHwMutexUnLock(wolfSSL_Mutex* mutex);
 
     #undef SHA_CTX
 
-    typedef enum {
-        ESP32_SHA_INIT = 0,
-        ESP32_SHA_HW = 1,
-        ESP32_SHA_SW = 2,
+    typedef enum
+    {
+        ESP32_SHA_INIT             = 0,
+        ESP32_SHA_HW               = 1,
+        ESP32_SHA_SW               = 2,
+        ESP32_SHA_HW_COPY          = 3,
         ESP32_SHA_FAIL_NEED_UNROLL = -1
     } ESP32_MODE;
 
-    typedef struct {
```

Comment:
Other than the  `isfirstblock`, all items in the struct were 4 bytes long, regardless of compiler optimization settings. The `isfirstblock` was already defined `byte` and listed last in sort order. I changed this to a `byte:1`, per the [code review suggestion below](https://github.com/wolfSSL/wolfssl/pull/6287#discussion_r1170587108). I also changed the `lockDepth` to a type of `byte:7`. If the depth ever exceeds even a relatively small number, things have gone seriously sideways. 

In the end, the items could have stayed in the same order, albeit a  smaller struct. I ended up moving the `initializer` pointer to the beginning, as intuitively `enum` _seems_ smaller, but is not (see below).

## Details

With a `bool` or `byte` type for `isfirstblock` , I suspect with a _reported_ size of `1`, there's likely a tiny amount of performance improvement when writing just one byte to update the `isfirstblock`  value, even though an additional three bytes appear to be wasted in memory in this case. The `bool` was _not_ interpreted by the compiler as `byte:1`.

### Analysis Setup

Inserting this code at the end of `esp_sha_init(WC_ESP32SHA* ctx ... )`
```c
    printf("Size of WC_ESP32SHA  %2d; address = %x\n", (int)sizeof(WC_ESP32SHA),       (unsigned int)ctx);

    printf("Size of mode         %2d; address = %x\n", (int)sizeof(ctx->mode),         (unsigned int)&ctx->mode);
    printf("Size of sha_type     %2d; address = %x\n", (int)sizeof(ctx->sha_type),     (unsigned int)&ctx->sha_type);
    printf("Size of initializer  %2d; address = %x\n", (int)sizeof(ctx->initializer),  (unsigned int)&ctx->initializer);
    printf("Size of lockDepth    %2d; address = %x\n", (int)sizeof(ctx->lockDepth),    (unsigned int)&ctx->lockDepth);
    printf("Size of isfirstblock %2d; address = %x\n", (int)sizeof(ctx->isfirstblock), (unsigned int)&ctx->isfirstblock);
```

I looked at two different compiler optimizations for the space actually used by the `WC_ESP32SHA`.

###  Optimize for Size

This was the setting with "optimize for size":

![image](https://user-images.githubusercontent.com/13059545/232993660-cfb204bf-4cf3-458f-a171-e7c1783fa5e2.png)


Here's the memory map of `ctx` just before the `return ret` in my [esp32_sha.c](https://github.com/gojimmypi/wolfssl/blob/ED25519_SHA2_fix/wolfcrypt/src/port/Espressif/esp32_sha.c#L84):

![image](https://user-images.githubusercontent.com/13059545/232992633-532e4357-f654-431b-8f5b-0ec02538354a.png)

Note in particular even though the `sizeof()` of the bool `ctx->isfirstblock` _claims_ to be 1, there are still 4 bytes used in memory and the total size of the `WC_ESP32SHA ctx` is 20 bytes (5 elements * 4 bytes)

### Optimize for Performance

I saw the same results for "optimize for performance". 

![image](https://user-images.githubusercontent.com/13059545/233002737-e134c31e-81c9-4201-a4b9-d05718d7d32c.png)

With this output and memory map:

![image](https://user-images.githubusercontent.com/13059545/232994735-2412f907-3e35-46db-8901-89ad6ddc198d.png)

### Adding an experimental additional boolean var to the struct

When adding an _additional_ boolean, we can see there's indeed a memory benefit in that the second value is placed in the same 4 byte memory block, at address 3ffc104d:

![image](https://user-images.githubusercontent.com/13059545/233001028-3371169e-60fc-4ee9-8db5-e4d7c6a234e5.png)

### Placing items out of size order

Not applicable to the final size in our case with only one boolean `isfirstblock` value, here's an example of detrimental memory allocation when one of the boolean variables are listed first, and the other is listed last.

![image](https://user-images.githubusercontent.com/13059545/233001546-8e958c38-c2d0-4d18-ad0b-c4efa0e54cd3.png)

Note there's a net increase of 4 bytes, with a total of 6 bytes now wasted. This is informative only and not applicable to this topic with only one `bool` in `WC_ESP32SHA`.

### Other Observations

It is disappointing the enums take up 4 bytes, such as the [Espressif SHA_TYPE](https://github.com/espressif/esp-idf/blob/1cb31e50943bb757966ca91ed7f4852692a5b0ed/components/esp_rom/include/esp32/rom/sha.h#L38).

I tried to force the wolfSSL `ESP32_MODE` enum to be only a single byte like this:

```
    typedef enum
    {
        ESP32_SHA_INIT             = (int8_t)0,
        ESP32_SHA_HW               = (int8_t)1,
        ESP32_SHA_SW               = (int8_t)2,
        ESP32_SHA_HW_COPY          = (int8_t)3,
        ESP32_SHA_FAIL_NEED_UNROLL = (int8_t)(-1)
    } ESP32_MODE;
```

However, the size was still 4 bytes. I was reminded by [stackoverflow](https://stackoverflow.com/questions/366017/what-is-the-size-of-an-enum-in-c) that from the [C Standard (C99)](http://www.open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf): 

>6.7.2.2 Enumeration specifiers
>[...]
>Constraints
>The expression that defines the value of an enumeration constant shall be an integer constant expression that has a value representable as an int.
>[...]
>Each enumerated type shall be compatible with char, a signed integer type, or an unsigned integer type. The choice of type is implementation-defined, but shall be capable of representing the values of all the members of the enumeration.


It would seem for the readability and type-checking convenience of enums, there's simply always a 3 byte memory penalty.

### Conclusion:

The `WC_ESP32SHA` is now only 16 bytes in size.

![image](https://user-images.githubusercontent.com/13059545/233031934-feed2721-a378-4112-a428-5e1760dc3043.png)